demos/time_dependent_rayleigh_benard/timedep-rayleigh-benard.py.rst

```diff
@@ -0,0 +1,186 @@
+difference is applied horizontally, perpendicular to gravity.  It
+tends to make prettier pictures for low Rayleigh numbers, but also
+tends to take more Newton iterations since the coupling terms in the
+Jacobian are a bit stronger.  Switching to the first case would be a
+simple change of the boundary subdomains associated with the second and
+third boundary conditions below::
+
+  bcs = [
+      DirichletBC(Z.sub(0), Constant((0, 0)), (1, 2, 3, 4)),
+      DirichletBC(Z.sub(2), Constant(1.0), (3,)),
+      DirichletBC(Z.sub(2), Constant(0.0), (4,))
+  ]
+
+Like Navier-Stokes, the pressure is only defined up to a constant.::
+
+  nullspace = [(1, VectorSpaceBasis(constant=True), 1)]
+
+Set up the Butcher tableau to use for time-stepping::
+
+  num_stages = 2
+  butcher_tableau = RadauIIA(num_stages)
+  
+We are going to carry out time stepping via Irksome, but we need
+to say how to solve the rather interesting stage-coupled system.
+We will use an outer Newton method with linesearch.
+The linear solver will be flexible GMRES.  We adapt the the tolerance of
+the inner solver via the Eisenstant-Walker trick using `snes_ksp_ew`.
+The linear solver will be preconditioned with a multigrid method.
+As a relaxation scheme, we apply several iterations (accelerated via GMRES)
+of a Vanka-type patch smoother via `ASMVankaPC`.  This smoother sets up a sequence of local problems involving all degrees of freedom for each field for each
+Runge--Kutta stage on the cells containing a vertex in the mesh.
+We use `exclude_inds` to indicate that we use velocity degrees of freedom on
+the patch boundary but exclude the pressure and temperature degrees of freedom.
+::
+
+  exclude_inds = ",".join([str(3*i+j) for i in range(num_stages) for j in (1, 2)])
+
+  params = {
+      "mat_type": "aij",
+      "snes_type": "newtonls",
+      "snes_converged_reason": None,
+      "snes_linesearch_type": "l2",
+      "snes_monitor": None,
+      "ksp_type": "fgmres",
+      "ksp_monitor": None,
+      "ksp_max_it": 200,
+      "ksp_atol": 1.e-12,
+      "snes_rtol": 1.e-10,
+      "snes_atol": 1.e-10,
+      "snes_ksp_ew": None,
+      "pc_type": "mg",
+      "mg_levels": {
+          "ksp_type": "gmres",
+          "ksp_max_it": 3,
+          "ksp_convergence_test": "skip",
+          "pc_type": "python",
+          "pc_python_type": "firedrake.ASMVankaPC",
+          "pc_vanka_construct_dim": 0,
+	  "pc_vanka_backend_type": "tinyasm",
+          "pc_vanka_exclude_subspaces": exclude_inds},
+      "mg_coarse": {
+          "ksp_type": "preonly",
+          "pc_type": "lu",
+          "pc_factor_mat_solver_type": "mumps",
+          "mat_mumps_icntl_14": 200}
+  }
+
+  stepper = TimeStepper(F, butcher_tableau, t, dt, upT, bcs=bcs,
+                        nullspace=nullspace, solver_parameters=params)
+
+Now that the stepper is set up, let's run over many time steps::
+
+
+  plot_freq = 10
+  Ts = []
+  cur_step = 0
+  while float(t) < 1.0:
+      print(f"t = {float(t)}")
```

Comment:
If I had intended you to put print statements in time loops, I would never have given you `ProgressBar`

Reply:
This will be the only firedrake demo with a progress bar, FYI.